13_number_theory_basics/practice/src/prime_numbers.c

```diff
@@ -1,159 +1,170 @@
-        int64_t value = check_values[i]; // this value we will check in Prime test
-        t = clock(); // hi-res clock counter
-        ferma(value);
-        t = clock() - t;
-        double time_taken = ((double)t) / CLOCKS_PER_SEC; // in seconds
-        owerall_time_taken += time_taken;
-    }
-    printf("%d iterations, executed in %f seconds\n", N_ITERATIONS, owerall_time_taken);
-}
-
+#include <stdint.h>
+#include <time.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdbool.h>
+
+// https://blog.skillfactory.ru/nok-i-nod-chisel-chto-eto-takoe-i-kak-ih-nayti/
+int64_t gcd(int64_t a, int64_t b){ // Euclidean algorithm
+    if (b == 0) {
+        return a;
+    }
+    return gcd(b, a%b);
+}
+
+int64_t lcm(int64_t a, int64_t b){
+    return (a * b) / gcd(a, b);
+}
+
+// https://ru.wikipedia.org/wiki/Решето_Эратосфена
+// need to call free
+bool* sieve_of_eratosthenes(int64_t size) {
+    bool* sieve = (bool*)malloc(size * sizeof(bool));
+    for (int64_t i = 0; i < size; ++i) {
+        sieve[i] = false;
+    }
+    sieve[1] = true;
+    for (int64_t i = 2; i < size; ++i) {
+        if (!sieve[i]) {
+            for (int64_t j = 2 * i; j < size; j += i) {
+                sieve[j] = true;
+            }
+        }
+    }
+    return sieve;
+}
+
+bool is_prime_naive(int64_t x) {
+    for (int64_t i = 2; i < x; ++i) {
+        if (x % i == 0) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool is_prime_sqrt(int64_t x) {
+    for (int64_t i = 2; i * i <= x; ++i) {
+        if (x % i == 0) {
+            return false;
+        }
+    }
+    return true;
+}
+
+// https://habr.com/ru/companies/otus/articles/779396/
+// бинарный алгоритм возведения в степень
+int64_t bin_pow(int64_t base, int64_t p) {
+    int64_t res = 1;
+    while (p) {
+        if (p & 1) {
+            res = (res * base);
+        }
+        base = (base * base);
+        p /= 2;
+    }
+    return res;
+}
+
+int64_t bin_pow_modulo(int64_t base, int64_t p, int64_t mod) {
+    int64_t res = 1;
+    while (p) {
+        if (p & 1) {
+            res = (res * base) % mod;
+        }
+        base = (base * base) % mod;
+        p /= 2;
+    }
+    return res;
+}
+#define FERMA_ITERATIONS 100
+
+// https://ru.wikipedia.org/wiki/Тест_Ферма
+bool ferma(int64_t x){
+    if (x == 2) {
+        return true;
+    }
+    for (int i = 0; i < FERMA_ITERATIONS; i++) {
+        int64_t a = (rand() % (x - 2)) + 2; // Pick a random numver [2..n-2]
+        if (gcd(a, x) != 1) {
+            return false;
+        }
+        if (bin_pow_modulo(a, x - 1, x) != 1) {
+            return false;
+        }
+    }
+    return true;
+}
+
+#define N_ITERATIONS 10000
+#define NUMBER_LIMITS 100000
+int main()
+{
+    srand(time(NULL));
+    clock_t t;
+    double owerall_time_taken = 0;
+    printf("Started naive prime checker\n");
+    printf("Onde iteration is generate random number in range 0-%d and check is it prime or not\n", NUMBER_LIMITS);

```

Comment:
I suppose it's more clear if range set as in the suggestion
```suggestion
    printf("Onde iteration is generate random number in range [0, %d] and check is it prime or not\n", NUMBER_LIMITS);

```